lib/Dialect/Linalg/Transforms/LinalgTransforms.cpp

```diff
@@ -106,3 +111,38 @@ bool mlir::linalg::detail::isProducedByOpOfTypeImpl(
   }
   return false;
 }
+
+LogicalResult
+mlir::linalg::permuteGenericLinalgOp(PatternRewriter &rewriter, Operation *op,
+                                     ArrayRef<unsigned> permutation,
+                                     StringRef linalgMarker) {
+  // If permutation is empty, there is nothing to be done.
+  if (permutation.size() == 0)
+    return failure();
+
+  OpBuilder::InsertionGuard g(rewriter);
+  auto linOp = cast<LinalgOp>(op);
+  rewriter.setInsertionPoint(linOp);
+  ScopedContext scope(rewriter, linOp.getLoc());
+  auto permutationMap =
+      inversePermutation(AffineMap::getPermutationMap(permutation, scope.getContext()));
+  SmallVector<Attribute, 4> newIndexingMap;
+  auto indexingMaps = linOp.getAttrOfType<ArrayAttr>("indexing_maps").getValue();
+  for (unsigned i = 0, e = linOp.getNumInputsAndOutputs(); i != e; ++i) {
+    AffineMap m = indexingMaps[i].cast<AffineMapAttr>().getValue().compose(permutationMap);
+    newIndexingMap.push_back(AffineMapAttr::get(m));
+  }
+  auto itTypes = linOp.getAttrOfType<ArrayAttr>("iterator_types").getValue();
+  SmallVector<Attribute, 4> itTypesVector(itTypes.begin(), itTypes.end());
+  applyPermutationToVector(itTypesVector, permutation);
```

Comment:
@River707 added the ValueRange class that avoids the need for this.
Can you maybe use this and update in the few places that will require it?

See https://github.com/tensorflow/mlir/pull/305/files for example updates.

Reply:
Problem here is that `applyPermutationToVector` actually requires a SmallVector<T,N> as its first argument so non-`Value` vectors can be permuted. If we switch to `ValueRange` here we will need a specific non-templated implementation that cannot be used for the single other use of this function.

Reply:
ah yes, thanks1